### ISSUE TITLE
S3 support

### ISSUE THREAD

**Issue by krakazyabra** [filebrowser/filebrowser#1077]
Hi! It will be nice, if here will be S3 support (not only AWS, but any custom S3 endpoint, like local Minio, ceph etc)

**Comment by garethgeorge** [comment]
I'd also be interested in this and can possibly contribute some time PR the feature. Can anyone on the project comment architecturally on how hard this would be to do / what needs to be done?

-- EDIT -- 

it looks like the project is using https://github.com/spf13/afero as its fs abstraction, if S3 support were contributed to that project would maintainers here consider bringing support?

**Comment by 719media** [comment]
Just trying to help here:

You are going to run into a problem really quick when trying to support s3. The main problem is that with s3, there really is not efficient "list directory" option, and any half-baked "workaround" gets REALLY BAD at scale. The way most people solve this is by using a database to store information about the files in s3, which I imagine falls way outside the scope of support you want to create with your repo.

For example, even though this article is 5 years old, the limitations it explains about accessing s3 as a "file browser" still apply: https://dzone.com/articles/building-a-file-explorer-on-top-of-amazon-s3

In short, in s3, you can only query ALL of your files at once, sorted in binary order, optionally matching a prefix. No sorting by reverse order, last modified, folders first, etc.

Other complications include moving a "folder" (renaming as well) can be a complex operation as you are actually copying and deleting EVERY FILE within that "folder".

**Comment by nambrosch** [comment]
i haven't tried it yet but keeping files locally and executing the s3 command line client upon file operations might be an option for some scenarios.  The fact that operations can be delayed network-wide is a real snag though, for example i've seen deleting a file take several minutes to fully propagate.

**Comment by fdefilippo** [comment]
Hi, you could take a cue from what was done in the sftpgo project: 

https://github.com/drakkan/sftpgo/blob/master/vfs/s3fs.go

s3 is used as a backend file system.

ps: as workaround, on linux, just using https://github.com/s3fs-fuse/s3fs-fuse to mount the root directory.

**Comment by o1egl** [comment]
@krakazyabra cloud storage support is out of the scope of this project. As @fdefilippo mentioned you can mound your s3 bucket using fuse.

**Comment by filiphanes** [comment]
Hi, I am voting for S3 support, too.

Was looking for alternatives to https://github.com/mickael-kerjean/filestash.
Filebrowser has nicer frontend, Filestash has more features.

@719media listing files and folders can be done without listing all objects https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
Move/Rename folder operation is not efficient, but can be implemented as I have done for Filestash https://github.com/mickael-kerjean/filestash/pull/405
Yes, it is SLOW, in my tests it moved about 3 objects per second on Scaleway S3, but can be parallelized. Users should be aware that renaming folder with many files can be long operation.

But for personal/small use cases it is cheaper to mount S3 and run filebrowser on small VPS.

**Comment by KevinOBytes** [comment]
A work-around I've found when trying to get S3 to act as the storage back-end for Filebrowser, at least when deploying Filebrowser as a container, is to mount the Filebrowser storage folder to an NFS share interface tying back to S3.

You can find the directions for Mounting Amazon S3 to an Amazon EC2 instance using a private connection to S3 File Gateway [here.](https://aws.amazon.com/blogs/storage/mounting-amazon-s3-to-an-amazon-ec2-instance-using-a-private-connection-to-s3-file-gateway/) It demonstrate how to mount Amazon S3 as an NFS volume to an EC2 instance using private connections to [AWS Storage Gateway and S3 using VPC endpoints. From there just point your configuration to use the NFS location as the storage for Filebrowser.
